tiledb/sm/serialization/query.cc

```diff
@@ -265,89 +265,112 @@ Status subarray_to_capnp(
+          range_subset[i] = RangeSetAndSuperset(
+              dim->type(), dim->domain(), false, coalesce_ranges);
+        }
+        // Add custom ranges, clearing any implicit ranges previously set.
+        for (const auto& range : ranges) {
+          throw_if_not_ok(range_subset[i].add_range_unrestricted(range));
+        }
+      }
     } else {
       // Handle 1.7 style ranges where there is a single range with no sizes
+      auto data = range_reader.getBuffer();
+      auto data_ptr = data.asBytes();
       Range range(data_ptr.begin(), data.size());
-      RETURN_NOT_OK(subarray->set_ranges_for_dim(i, {range}));
-      subarray->set_is_default(i, range_reader.getHasDefaultRange());
+      throw_if_not_ok(range_subset[i].add_range_unrestricted(range));
     }
   }
 
+  std::vector<optional<Subarray::LabelRangeSubset>> label_range_subset(
+      dim_num, nullopt);
   if (reader.hasLabelRanges()) {
-    subarray->add_default_label_ranges(dim_num);
     auto label_ranges_reader = reader.getLabelRanges();
     uint32_t label_num = label_ranges_reader.size();
     for (uint32_t i = 0; i < label_num; i++) {
       auto label_range_reader = label_ranges_reader[i];
-      auto dim_id = label_range_reader.getDimensionId();
+      auto dim_index = label_range_reader.getDimensionId();
+      auto dim = array->array_schema_latest().dimension_ptr(dim_index);
       auto label_name = label_range_reader.getName();
 
       // Deserialize ranges for this dim label
       auto range_reader = label_range_reader.getRanges();
       auto ranges = range_buffers_from_capnp(range_reader);
 
       // Set ranges for this dim label on the subarray
-      subarray->set_label_ranges_for_dim(dim_id, label_name, ranges);
+      label_range_subset[dim_index] = {
+          label_name, dim->type(), coalesce_ranges};
     }
   }
 
+  std::unordered_map<std::string, std::vector<Range>> attr_range_subset;
   if (reader.hasAttributeRanges()) {
-    std::unordered_map<std::string, std::vector<Range>> attr_ranges;
     auto attr_ranges_reader = reader.getAttributeRanges();
     if (attr_ranges_reader.hasEntries()) {
-      for (auto attr_ranges_entry : attr_ranges_reader.getEntries()) {
-        auto range_reader = attr_ranges_entry.getValue();
-        auto key = std::string_view{
-            attr_ranges_entry.getKey().cStr(),
-            attr_ranges_entry.getKey().size()};
-        attr_ranges[std::string{key}] = range_buffers_from_capnp(range_reader);
+      for (auto entry : attr_ranges_reader.getEntries()) {
+        auto range_reader = entry.getValue();
+        std::string key{entry.getKey().cStr(), entry.getKey().size()};
+        attr_range_subset[key] = range_buffers_from_capnp(range_reader);
       }
     }
-
-    for (const auto& attr_range : attr_ranges)
-      subarray->set_attribute_ranges(attr_range.first, attr_range.second);
   }
 
   // If cap'n proto object has stats set it on c++ object
+  Subarray s(array, layout, parent_stats, logger, true);
```

Comment:
We need something better here... I'm working with @robertbindar to make something better. Your new constructor will need to take in a pointer to the parent stats as well as the deserialized stats (that will be contained in a new transition object).